### ISSUE TITLE
[Qwen3-TTS][Research] Device-resident Talker feedback ownership

### ISSUE THREAD

**Issue by charliechenye** [charliechenye/sglang-omni#25]
Parent roadmap: #19

Upstream context:
- sgl-project/sglang-omni#1754
- sgl-project/sglang-omni#1462 already removed major Qwen3-TTS AR host-path waste and replaced per-row output clones with batched snapshots

## Question

After the host-path cleanup already landed in #1462, is next-step Talker feedback still paying meaningful Python ownership, snapshot, allocation, and restaging overhead?

The remaining dataflow is conceptually:

```text
Code Predictor / Talker outputs on GPU
  -> batched output snapshot
  -> per-request scheduler-owned feedback object/queue
  -> next decode step iterates active requests
  -> feedback rows are assembled/staged again
  -> graph/static decode input buffer
```

The first-principles alternative is to keep the feedback dependency device-resident and indexed by stable request slots, so the next decode consumes the prior step's result without round-tripping through Python tensor ownership.

This issue is deliberately research-gated. Do not implement slot state until the current path is measured.

## Hypothesis

At moderate/high concurrency, the remaining feedback path may still be a material fraction of scheduler CPU or per-step exposed host time due to some combination of:

- batched `codes_snap` / `embeds_snap` clone/allocation
- per-request queue/list ownership
- request-order lookup/reassembly
- `torch.stack` or equivalent row assembly
- feedback staging/copy into the decode graph/static buffer
- cleanup/retraction bookkeeping

If material, a request-slot-indexed device-resident feedback pool may remove part of this path.

## Phase 0: attribution only

Instrument current upstream main without changing ownership.

Attribute per decode step:

- output snapshot clone time and bytes
- temporary allocation count/bytes
- Python time assigning feedback to requests
- pending feedback queue operations
- next-step feedback collection/reassembly time
- feedback H2D/D2D copy count and bytes
- scheduler thread CPU share
- exposed host idle gap between GPU steps

Profile at:

- c1 / c8 / c16 / c32
- 0.6B and 1.7B if the bottleneck differs
- normal seeded sampling settings
- CUDA Graph enabled production path

Use py-spy and/or profiler instrumentation that does not itself perturb the server into the known `with_stack` deadlock regime.

## Promotion gate for implementation

Only continue if the feedback ownership/staging cluster is a meaningful bottleneck.

Suggested trigger:

- >= 5% of scheduler/step time, or
- >= 5% of exposed host critical-path gap, or
- a clear allocation/copy hotspot whose isolated removal has a plausible end-to-end effect

The percentages are triage thresholds, not performance promises.

If the cluster is below the local noise floor, close the issue with the attribution result.

## Candidate design

If Phase 0 passes:

### Stable request slots

Maintain a bounded engine-owned slot pool:

```text
request_id -> slot_id + generation
```

Each slot owns the device-resident feedback needed for the next Talker decode.

### Producer path

Write Code Predictor/Talker feedback directly into the slot's device buffer or into a stable batched buffer whose rows map to slots.

Avoid creating a per-request GPU tensor merely to hand it back to the next GPU step.

### Consumer path

Build the next decode batch from slot indices/views rather than a Python list of feedback tensors.

The scheduler may still own logical request order; device state ownership should not require request order to equal slot order.

### Generation-safe reuse

Every slot reuse must carry a generation/version so a late completion or cancellation from request A cannot write into request B after the slot is recycled.

## Lifecycle correctness requirements

This path is stateful and therefore must explicitly test:

- request reorder
- batch shrink/growth
- finish
- EOS on one row while peers continue
- abort/cancel
- retraction
- re-prefill after retraction
- request-id reuse
- slot release
- generation-safe slot reuse
- exceptions during producer/consumer work
- server shutdown with in-flight requests

No request may read feedback produced by another request or by an earlier generation of the same slot.

## CUDA Graph requirements

- slot-backed state must have stable addresses where capture requires them
- graph replay must not capture transient request Python objects
- inactive bucket rows must be neutralized safely
- graph memory ownership and stream ordering must remain explicit
- deterministic-inference behavior must match the current reference path

## Benchmark requirements if implemented

Report before/after:

- scheduler CPU time
- per-step host gap
- snapshot/allocation counts
- D2D/H2D copy counts and bytes
- request throughput
- TTFA p50/p95
- total latency p50/p95
- inter-chunk p95/p99
- persistent GPU-memory overhead of the slot pool
- c1/c8/c16/c32

A microbenchmark improvement is not enough for a serving-performance claim.

## Stop conditions

Close without implementation if:

- the current batched snapshot/staging path is already negligible after #1462
- slots add more persistent memory/complexity than the measured host cost justifies
- async decode or another upstream change removes the ownership path first
- correctness requires coupling too deeply to generic SGLang request ordering for a small gain

## Non-goals

- changing Code Predictor numerics
- replacing the SGLang scheduler
- duplicating Codec state-slot work from upstream #1754
- adding async decode solely to justify this state design

## Deliverables

- [ ] Current feedback-path attribution
- [ ] Copy/allocation ownership diagram
- [ ] Go / no-go verdict
- [ ] If go: slot lifecycle/state-machine design
- [ ] Correctness tests for reorder/retract/finish/abort/reuse
- [ ] End-to-end A/B

**Comment by charliechenye** [comment]
## Retirement note

Closing this tracker as **not planned in its current form** because the upstream Talker feedback path changed materially after this issue was written.

The key upstream change is sgl-project/sglang-omni#1971, merged on 2026-09-07: **"Remove the dead work from the code predictor replay and batch the feedback write."**

That PR directly attacks the path this issue intended to profile/design around:

- removes 149 dead kernels from the default Code Predictor replay;
- stages sampling inputs more efficiently;
- replaces the previous per-row feedback write pattern with a batched feedback/text-row write;
- changes decode-input history ownership/retraction handling around the same feedback flow;
- reports about +3.4% QPS at c1 and +3.8% at c16 on its validated Qwen3-TTS workload.

The original #25 hypothesis was that the remaining post-#1462 feedback snapshot / per-request queue / row reassembly / restaging cluster might justify a request-slot-indexed device-resident feedback pool. After #1971, that hotspot description is no longer a trustworthy map of current main.

This closure does **not** claim that all feedback ownership overhead is now zero. It means we should not carry forward the proposed slot-pool architecture from stale evidence.

Only reopen this direction if a fresh current-main profile after #1971 shows a remaining feedback ownership/staging cluster that clears the original promotion gate (roughly >=5% scheduler/step time or exposed host critical-path gap). If such evidence appears, start a new narrowly scoped tracker from the new dataflow rather than reviving the old design.
